File: ts/packages/core/src/ports/storage.ts
// A minimal async key/value contract for whatever this package needs to persist (revocation entries, data-domain oplog entries) -- local and remote implementations interchangeable behind it, per the org's storage-boundary rules. Values are opaque bytes, not a specific serialisation: callers encode/decode with whatever schema the stored data actually has (a revocation-entry, a data-domain entry, etc.), matching the protocol's own treatment of data-domain entries as opaque to the wire format.
export interface KeyValueStorage {
  get: (key: string) => Promise<Uint8Array | undefined>;
  set: (key: string, value: Uint8Array) => Promise<void>;
  delete: (key: string) => Promise<void>;
  /** Keys with the given prefix, in no particular order. */
  keys: (prefix: string) => Promise<string[]>;
}

File: rust/crates/wire-mesh-core/src/ports/storage.rs
//! The storage port. Values are opaque bytes — a revocation entry, an
//! oplog entry; callers encode and decode with the wire types, exactly as
//! the protocol's own data domain treats its entries.

use crate::ports::CoreError;

/// Key-value storage with prefix enumeration, mirroring the TypeScript
/// core's storage contract.
#[async_trait::async_trait]
pub trait KeyValueStorage: Send + Sync {
    async fn get(&self, key: &str) -> Result<Option<Vec<u8>>, CoreError>;
    async fn set(&self, key: &str, value: Vec<u8>) -> Result<(), CoreError>;
    async fn delete(&self, key: &str) -> Result<(), CoreError>;
    /// All keys starting with `prefix`, in lexicographic order.
    async fn keys(&self, prefix: &str) -> Result<Vec<String>, CoreError>;
}
